jetty-io/src/main/java/org/eclipse/jetty/io/Connection.java

```diff
@@ -96,31 +96,49 @@
 
     long getCreatedTimeStamp();
 
+    /**
+     * <p>{@link Connection} implementations implement this interface when they
+     * can upgrade from the protocol they speak (for example HTTP/1.1)
+     * to a different protocol (e.g. HTTP/2).</p>
+     *
+     * @see EndPoint#upgrade(Connection)
+     * @see UpgradeTo
+     */
     interface UpgradeFrom
     {
         /**
-         * <p>Takes the input buffer from the connection on upgrade.</p>
-         * <p>This method is used to take any unconsumed input from
-         * a connection during an upgrade.</p>
+         * <p>Invoked during an {@link EndPoint#upgrade(Connection) upgrade}
+         * to produce a buffer containing bytes that have not been consumed by
+         * this connection, and that must be consumed by the upgrade-to
+         * connection.</p>
          *
-         * @return A buffer of unconsumed input. The caller must return the buffer
-         * to the bufferpool when consumed and this connection must not.
+         * @return a buffer of unconsumed bytes to pass to the upgrade-to connection.
+         * The buffer does not belong to any pool and should be discarded after
+         * having consumed its bytes.
+         * The returned buffer may be null if there are no unconsumed bytes.
          */
         ByteBuffer onUpgradeFrom();
     }
 
+    /**
+     * <p>{@link Connection} implementations implement this interface when they
+     * can be upgraded to the protocol they speak (e.g. HTTP/2)
+     * from a different protocol (e.g. HTTP/1.1).</p>
+     */
     interface UpgradeTo
     {
         /**
-         * <p>Callback method invoked when this connection is upgraded.</p>
-         * <p>This must be called before {@link #onOpen()}.</p>
+         * <p>Invoked during an {@link EndPoint#upgrade(Connection) upgrade}
+         * to receive a buffer containing bytes that have not been consumed by
+         * the upgrade-from connection, and that must be consumed by this
+         * connection.</p>
          *
-         * @param prefilled An optional buffer that can contain prefilled data. Typically this
-         * results from an upgrade of one protocol to the other where the old connection has buffered
-         * data destined for the new connection.  The new connection must take ownership of the buffer
-         * and is responsible for returning it to the buffer pool
+         * @param buffer a non-null buffer of unconsumed bytes received from
+         * the upgrade-from connection.
+         * The buffer does not belong to any pool and should be discarded after
+         * having consumed its bytes.
          */
-        void onUpgradeTo(ByteBuffer prefilled);
+        void onUpgradeTo(ByteBuffer buffer);
     }
```

Comment:
The contract changes you're proposing forces a `ByteBuffer` allocation.

Couldn't that be avoided by specifying that the buffer returned by `onUpgradeFrom` and passed to `onUpgradeTo` must be fully consumed and no reference to it should be kept?

I would phrase that as:

For `onUpgradeTo`: Implementations MUST consume all bytes of the buffer and MUST NOT keep a reference to the buffer.
For `onUpgradeFrom`: The buffer WILL only be used for the duration of the `Endpoint.upgrade` call and MAY be disposed of afterwards.

Reply:
@lorban and I discussed this and the problem with the approach he proposes is that `onUpgradeFrom()` returns a buffer, but it won't know when it's free to release it back to the pool.